### PR TITLE
Replace the "Status of the Python branches" table with a csv.

### LINKED ISSUE #883: Provide the status of Python branches and versions as a csv
**Describe the enhancement or feature you'd like**
Currently we have to open a browser and read the status of Python branches.

https://devguide.python.org/#status-of-python-branches

It would be great if that info is available in CSV format instead of in an rst table.

If the data is in CSV, we can then write scripts to programmatically parse the table and find the info we need.

According to @ezio-melotti , Sphinx support the `.. csv-table::` directive, so we can use that instead of the current rst table syntax.

### PR DESCRIPTION
**Pull request by ezio-melotti**
This PR fixes #883.

I kept the Sphinx markup in the csv, or otherwise it would have been lost.  The markup should be easy enough to remove while parsing, and it only affects PEPs and dates.  I checked if there was a way to include a separate column without markup and then ignore it in the table, but there's no such option.

While I was at it, I also fixed the column widths.  Here's a comparison of the result (csv table on top, old one below):

image

### CODE REVIEW

**Review by Mariatta — APPROVED** [review]
The csv table looks much cleaner this way.

**Comment by encukou** [comment]
Could you convert the EOL table as well? https://devguide.python.org/devcycle/#end-of-life-branches

**Comment by ezio-melotti** [comment]
If we start adding more csv files, I wonder if it would make sense to add them in a new `include` dir.  The reasons for doing this are:
* if people start using the files now, and then we move them, we will break their scripts;
* the root dir is already crowded with rst (and other files);
* I'm planning to add more files in #848.

**Comment by ezio-melotti** [comment]
Relevant discussion about the two tables (cc @hugovk, @encukou): https://github.com/endoflife-date/endoflife.date/discussions/711

**Comment by ezio-melotti** [comment]
I updated the PR to include the end-of-life table, and moved both the `csv`s in a new `include` directory.
I'll wait after #901 is done to merge this, as to not create additional conflicts in that PR.

**Comment by AA-Turner** [comment]
Does `auto` work for widths?

Another option would be to combine the CSVs, and use a very simple extension in `conf.py` to split them by EOL/active and add the italics -- that would allow for easier parsing for external consumers such as endoflife.date. 

A

**Comment by ezio-melotti** [comment]
> Does auto work for widths?

I can give that a try.  I set those widths manually in order to keep each row in a single line (see e.g. the screenshots in the first message).

> Another option would be to combine the CSVs,

I'm trying to gather some feedback from the endoflife folks and pydotorg in order to figure out their needs.  If it turns out that a single table is better, I can combine them.

> and use a very simple extension in conf.py to split them by EOL/active and add the italics -- that would allow for easier parsing for external consumers such as endoflife.date.

What extension?  I also don't like too much the rst markup in the csv, especially the `` :pep:`...` ``s applied to all the PEPs in the PEP column, but the italic is not applied to all rows and it's easy enough to remove afterwards that it's probably better to keep it there.

**Comment by ezio-melotti** [comment]
Hmm, locally the tables look fine after removing the `:widths:`, but on https://cpython-devguide--884.org.readthedocs.build/#status-of-python-branches they don't.  Here I tested with both Sphinx 5.0.1 and 5.0.2 -- perhaps readthedocs is using an older version?

**Comment by CAM-Gerlach** [comment]
For reference, this is how they look in FF on Windows on the preview.

image

> I checked if there was a way to include a separate column without markup and then ignore it in the table, but there's no such option.

You could use [Sphinx CSV Filter](https://github.com/crate/sphinx_csv_filter), both to include/exclude specific columns (e.g. with/without markup), and also to only include specific roles (e.g. that are marked EOL or not). That would avoid having to write a custom directive.

**Comment by ezio-melotti** [comment]
I updated the PR after the Devguide reorganization, reintroduced the `:widths:` to fix the rendering issues and also to make both tables consistent, and fixed a few other minor things.

I'm not sure if it's worth unifying the two CSVs and/or tables, since that adds an extra dependency and complexity.

**Comment by CAM-Gerlach** [comment]
> I'm not sure if it's worth unifying the two CSVs and/or tables, since that adds an extra dependency and complexity.

You could just put them in the same table, since there's already a `Status` column, if there were an easy way to clearly separate supported from unsupported statuses (e.g. by row color). But I'm not sure if there is.

File: include/branches.csv [hunk 1]
```diff
@@ -0,0 +1,7 @@
+Branch,Schedule,Status,First release,End-of-life,Release manager
+main,TBA,features,*2023-10-03*,*2028-10*,Thomas Wouters
```
Comment (hugovk) [inline include/branches.csv:2]:
```suggestion
main,:pep:`693`,features,*2023-10-03*,*2028-10*,Thomas Wouters
```

**Comment by hugovk** [comment]
> I'll wait after #901 is done to merge this, as to not create additional conflicts in that PR.

#901 is now merged.

**Review by hugovk — APPROVED** [review]
It would be useful to have machine-readable data, for example to automate https://python-release-cycle.glitch.me/

JSON would be preferred but CSV should be fine too.

(https://endoflife.date/api/python.json doesn't do prereleases.)

File: include/branches.csv [hunk 2]
```diff
@@ -0,0 +1,7 @@
+Branch,Schedule,Status,First release,End-of-life,Release manager
+main,:pep:`693`,features,*2023-10-03*,*2028-10*,Thomas Wouters
+3.11,:pep:`664`,bugfix,*2022-10-03*,*2027-10*,Pablo Galindo Salgado
```
Comment (CAM-Gerlach) [inline include/branches.csv:3]:
```suggestion
3.11,:pep:`664`,bugfix,*2022-10-24*,*2027-10*,Pablo Galindo Salgado
```

Fix the actual release date

File: include/branches.csv [hunk 3]
```diff
@@ -0,0 +1,7 @@
+Branch,Schedule,Status,First release,End-of-life,Release manager
+main,:pep:`693`,features,*2023-10-03*,*2028-10*,Thomas Wouters
```
Comment (CAM-Gerlach) [inline include/branches.csv:2]:
```suggestion
main,:pep:`693`,features,*2023-10-02*,*2028-10*,Thomas Wouters
```

Use the release data specified in PEP 693

**Review by CAM-Gerlach — COMMENTED** [review]
A couple updates to the release dates

File: include/branches.csv [hunk 4]
```diff
@@ -0,0 +1,7 @@
+Branch,Schedule,Status,First release,End-of-life,Release manager
+main,:pep:`693`,features,*2023-10-02*,*2028-10*,Thomas Wouters
+3.11,:pep:`664`,bugfix,*2022-10-24*,*2027-10*,Pablo Galindo Salgado
```
Comment (ezio-melotti) [inline include/branches.csv:3]:
```suggestion
3.11,:pep:`664`,bugfix,2022-10-24,*2027-10*,Pablo Galindo Salgado
```

**Comment by ezio-melotti** [comment]
As discussed in the docs meeting, I now merged this.  There are more changes we are considering, e.g. merging the two table or creating a JSON file that is then used to create these CSVs, but those will be handled in separate PRs.

**Comment by encukou** [comment]
FWIW, I think the source should be a human-writable format, and we should *generate* a JSON for the machines to consume (like https://peps.python.org/api/peps.json). And we should tell people to expect the source format to change whenever we feel like it.
(That'll need a Sphinx plugin, but if we want to move the formatting out of the tables, one will be needed anyway.)

**Comment by hugovk** [comment]
Thanks! Let's continue on https://github.com/python/docs-community/issues/67.

**Pull request merged**